### PR TITLE
Port Drop on Slip

### PR DESCRIPTION
**Pull request by Finket**
# Description

A Parkstation list item from https://github.com/Simple-Station/Parkstation-Friendly-Chainsaw/issues/2
I have not made any changes to the code, it might be worth looking at the clothing YAMLs to see if anything should be added or changed.

---
# Todo

- [ ] Add end of round stat when https://github.com/Simple-Station/Parkstation-Friendly-Chainsaw/pull/50 is merged

---

# Changelog

:cl:
- add: Added chance to drop certain items when slipped

### CODE REVIEW

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:1]: An end of round stat for items dropped would be funny

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly InventorySystem _invSystem = default!;
+    [Dependency] private readonly SharedPhysicsSystem _physics = default!;
+    [Dependency] private readonly ThrowingSystem _throwing = default!;
+
+    private static readonly float PocketDropChance = 10f;
+    private static readonly float PocketThrowChance = 5f;
+
+    private static readonly float ClumsyDropChance = 5f;
+    private static readonly float ClumsyThrowChance = 90f;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:29]:
```suggestion
    public const float PocketDropChance = 10f;
    public const float PocketThrowChance = 5f;

    public const float ClumsyDropChance = 5f;
    public const float ClumsyThrowChance = 90f;
```

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:30]:
```suggestion


```

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        if (!_invSystem.TryGetSlots(entity, out var slotDefinitions))
+            return;
+
+        foreach (var slot in slotDefinitions)
+        {
+            if (!_invSystem.TryGetSlotEntity(entity, slot.Name, out var item))
+                continue;
+
+            // A check for DropOnSlipComponent.
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:49]:
```suggestion
            // Check for DropOnSlipComponent
```

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            if (slot.Name != "pocket1" && slot.Name != "pocket2" && EntityManager.TryGetComponent<DropOnSlipComponent>(item, out var dropComp) && _random.NextFloat(0, 100) < dropComp.Chance)
+            {
+                var popupString = Loc.GetString("system-drop-on-slip-text-component", ("name", entity), ("item", item));
+
+                Drop(entity, item.Value, slot.Name, popupString);
+                continue;
+            }
+
+            // A check for any items in pockets.
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:58]:
```suggestion
            // Check for any items in pockets
```

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            if (slot.Name == "pocket1" | slot.Name == "pocket2" && _random.NextFloat(0, 100) < PocketDropChance)
+            {
+                var popupString = Loc.GetString("system-drop-on-slip-text-pocket", ("name", entity), ("item", item));
+
+                Drop(entity, item.Value, slot.Name, popupString);
+                continue;
+            }
+
+            // A check for ClumsyComponent.
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:67]:
```suggestion
            // Check for ClumsyComponent
```

File: Content.Server/SimpleStation14/Slippery/DropOnSlipComponent.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+[RegisterComponent]
+public sealed partial class DropOnSlipComponent : Component
+{
+    [DataField("chance")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public int Chance = 20;
+
+    [DataField("chanceToThrow")]
+    [ViewVariables(VVAccess.ReadWrite)]
+    public int ChanceToThrow = 40;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipComponent.cs:15]: I don't think this is used in code

Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipComponent.cs:15]:
```suggestion
    /// <summary>
    ///     Percent chance to drop this item when slipping
    /// </summary>
    [DataField("chance")]
    [ViewVariables(VVAccess.ReadWrite)]
    public int Chance = 20;

    /// <summary>
    ///     Percent chance to throw the item after successfully rolling to drop it
    /// </summary>
    [DataField("chanceToThrow")]
    [ViewVariables(VVAccess.ReadWrite)]
    public int ChanceToThrow = 40;
```

File: Resources/Prototypes/Entities/Clothing/Head/hats.yml [hunk 7]
```diff
@@ -40,17 +44,21 @@
...
-    - type: Sprite
-      sprite: Clothing/Head/Hats/casa.rsi
-    - type: Clothing
-      sprite: Clothing/Head/Hats/casa.rsi
+  - type: Sprite
+    sprite: Clothing/Head/Hats/casa.rsi
+  - type: Clothing
+    sprite: Clothing/Head/Hats/casa.rsi
+  - type: DropOnSlip
+    chance: 20
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Entities/Clothing/Head/hats.yml:61]:
```suggestion
    - type: Sprite
      sprite: Clothing/Head/Hats/casa.rsi
    - type: Clothing
      sprite: Clothing/Head/Hats/casa.rsi
    - type: DropOnSlip
      chance: 20
```

File: Resources/Prototypes/Entities/Clothing/Head/hats.yml [hunk 8]
```diff
@@ -929,7 +1025,7 @@
     - type: AddAccentClothing
       accent: ReplacementAccent
       replacement: cowboy
-      
+
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Entities/Clothing/Head/hats.yml:1028]:
```suggestion
      
```

File: Resources/Prototypes/Entities/Clothing/Head/hats.yml [hunk 9]
```diff
@@ -939,7 +1035,7 @@
       sprite: Clothing/Head/Hats/cowboyhatblack.rsi
     - type: Clothing
       sprite: Clothing/Head/Hats/cowboyhatblack.rsi
-      
+
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Entities/Clothing/Head/hats.yml:1038]:
```suggestion
      
```

File: Resources/Prototypes/Entities/Clothing/Head/hats.yml [hunk 10]
```diff
@@ -949,7 +1045,7 @@
       sprite: Clothing/Head/Hats/cowboyhatgrey.rsi
     - type: Clothing
       sprite: Clothing/Head/Hats/cowboyhatgrey.rsi
-      
+
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Entities/Clothing/Head/hats.yml:1048]:
```suggestion
      
```

File: Resources/Prototypes/Entities/Clothing/Head/hats.yml [hunk 11]
```diff
@@ -959,7 +1055,7 @@
       sprite: Clothing/Head/Hats/cowboyhatred.rsi
     - type: Clothing
       sprite: Clothing/Head/Hats/cowboyhatred.rsi
-      
+
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Entities/Clothing/Head/hats.yml:1058]:
```suggestion
      
```

File: Resources/Prototypes/Entities/Clothing/Head/hats.yml [hunk 12]
```diff
@@ -969,7 +1065,7 @@
       sprite: Clothing/Head/Hats/cowboyhatwhite.rsi
     - type: Clothing
       sprite: Clothing/Head/Hats/cowboyhatwhite.rsi
-      
+
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Entities/Clothing/Head/hats.yml:1068]:
```suggestion
      
```

File: Resources/Prototypes/Entities/Clothing/Head/misc.yml [hunk 13]
```diff
@@ -183,7 +193,7 @@
     sprite: Clothing/Head/Hats/catears.rsi
   - type: AddAccentClothing
     accent: OwOAccent
-    
+
```
Comment (DEATHB4DEFEAT) [inline Resources/Prototypes/Entities/Clothing/Head/misc.yml:196]:
```suggestion
    
```

**Review by DEATHB4DEFEAT — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Entities/Clothing/Head/hats.yml (hunk 7 above)
Comment (Finket) [reply Resources/Prototypes/Entities/Clothing/Head/hats.yml:61]: Was this a mistake? It's spaced like every other entity in this file

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+using Content.Shared.Throwing;
+
+namespace Content.Server.SimpleStation14.Slippery;
+
+public sealed class DropOnSlipSystem : EntitySystem
+{
+    [Dependency] private readonly ISharedAdminLogManager _adminLogger = default!;
+    [Dependency] private readonly PopupSystem _popup = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly InventorySystem _invSystem = default!;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:21]:
```suggestion
    [Dependency] private readonly InventorySystem _inventory = default!;
```

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<InventoryComponent, ParkSlipEvent>(HandleSlip);
+    }
+
+
+    private void HandleSlip(EntityUid entity, InventoryComponent invComp, ParkSlipEvent args)
+    {
+        if (!_invSystem.TryGetSlots(entity, out var slotDefinitions))
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:42]:
```suggestion
        if (!_inventory.TryGetSlots(entity, out var slotDefinitions))
```

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+            return;
+
+        foreach (var slot in slotDefinitions)
+        {
+            if (!_invSystem.TryGetSlotEntity(entity, slot.Name, out var item))
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:47]:
```suggestion
            if (!_inventory.TryGetSlotEntity(entity, slot.Name, out var item))
```

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+                Drop(entity, item.Value, slot.Name, popupString);
+                continue;
+            }
+        }
+    }
+
+    private void Drop(EntityUid entity, EntityUid item, string slot, string popupString)
+    {
+        if (!_invSystem.TryUnequip(entity, slot, false, true))
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:81]:
```suggestion
        if (!_inventory.TryUnequip(entity, slot, false, true))
```

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly InventorySystem _invSystem = default!;
+    [Dependency] private readonly SharedPhysicsSystem _physics = default!;
+    [Dependency] private readonly ThrowingSystem _throwing = default!;
+
+    public const float PocketDropChance = 10f;
+    public const float PocketThrowChance = 5f;
+
+    public const float ClumsyDropChance = 5f;
+    public const float ClumsyThrowChance = 90f;
```
Comment (DEATHB4DEFEAT) [inline Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:29]: Did I comment on these already? They probably shouldn't be constants.

**Review by DEATHB4DEFEAT — CHANGES_REQUESTED** [review]

File: Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs
Comment (Finket) [reply Content.Server/SimpleStation14/Slippery/DropOnSlipSystem.cs:1]: I placed this in todo and mentioned the end of round stats pr. Resolving comment.

**Review by DEATHB4DEFEAT — APPROVED** [review]

**Review by Pspritechologist — CHANGES_REQUESTED** [review]
This shouldn't be using a hardcoded method to check for pockets, etc., and overall needs to be done in a better way. Probably put off merging this until I give a pass of handling it differently.

File: Content.Server/SimpleStation14/Slippery/DropOnSlipComponent.cs [hunk 19]
```diff
@@ -0,0 +1,15 @@
+﻿namespace Content.Server.SimpleStation14.Slippery;
+
+/// <summary>
+///     Uses provided chance to try and drop the item when slipped, if equipped.
```
[the remark this answers is not in the document]
Comment (Finket) [reply Content.Server/SimpleStation14/Slippery/DropOnSlipComponent.cs]: Currently looking at https://www.github.com/space-wizards/space-station-14/issues/9702 as the solution.

**Pull request merged**